R/utils.R

```diff
@@ -55,3 +92,89 @@ str_standardise <- function(name) {
 str_trim_ext <- function(path) {
   sub("\\..+$", "", path)
 }
+
+gather_tree <- function(x, depth) {
+  if (is.character(depth)) {
+    depth <- as.list(depth)
+  } else if (is.numeric(depth)) {
+    depth <- rep(list(NULL), depth)
+  } else {
+    stop("`depth` should be character or numeric", call. = FALSE)
+  }
+
+  if (length(depth) > 1) {
+    x <- gather_tree_recurse(x, depth)
+  }
+  bind_rows(x, .id = depth[[1]])
+}
+
+gather_tree_recurse <- function(x, depth) {
+  if (length(depth) > 1) {
+    lapply(x, gather_tree_recurse, depth = depth[-1])
+  } else {
+    bind_rows(x, .id = depth[[1]])
+  }
+}
+
+at_bottom <- function(.x, .f, ...) {
+  if (is.list(.x)) {
+    lapply(.x, at_bottom, .f, ...)
+  } else {
+    .f(.x, ...)
+  }
+}
+
+spread_attributes <- function(x, id) {
+  stopifnot(is.atomic(x))
+
+  attrs <- attributes(unclass(x))
+  stopifnot(all(lengths(attrs) == length(x)))
+
+  attrs <- as.data.frame(attrs, stringsAsFactors = FALSE)
+  df <- cbind(unclass(x), attrs, stringsAsFactors = FALSE)
+  names(df)[[1]] <- id
+
+  df
+}
+
+# Simple reimplementations to avoid dplyr dependency -----------------
+
+bind_rows <- function(x, .id = NULL) {
+  df <- do.call("rbind", unname(x))
+
+  if (!is.null(.id)) {
+    lengths <- vapply(x, nrow, numeric(1))
+    id <- rep(names(x), lengths)
+    df <- cbind(id, df, stringsAsFactors = FALSE)
+    names(df)[[1]] <- .id
+  }
+
+  df
+}
+
+filter <- function(df, ...) {
+  fs <- list(...)
+  stopifnot(all(vapply(fs, inherits, logical(1), "formula")))
+  fs <- lapply(fs, `[[`, 2)
+
+  bind_call <- function(lhs, rhs, fun = "&") {
+    call(fun, lhs, rhs)
+  }
+  pattern <- Reduce(bind_call, fs[-1], fs[[1]])
+  args <- list(df, pattern, substitute())
```

Comment:
I think you need `args$drop <- FALSE` here to be safe (and below in `filter_first()` you also need `drop = FALSE`)